### PR TITLE
Provide support for using multiple MAX31865 sensors

### PR DESCRIPTION
**Pull request by brvdboss**
## Description:

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest dev branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR.
  - [x] The code change is tested and works on Tasmota core ESP8266 V.2.7.3.2
  - [x] The code change is tested and works on core ESP32 V.1.12.2
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

Note; for the ESP32-test, I did do al the builds and those compile. Did not test on actual hardware yet.
Did test on actual hardware for ESP8266

This is the implementation of what was discussed in https://github.com/arendst/Tasmota/issues/9086

### CODE REVIEW

File: tasmota/language/bg_BG.h [hunk 1]
```diff
@@ -622,6 +622,11 @@
 #define D_SENSOR_MAX31855_CS   "MX31855 CS"
 #define D_SENSOR_MAX31855_CLK  "MX31855 CLK"
 #define D_SENSOR_MAX31855_DO   "MX31855 DO"
+#define D_SENSOR_MAX31865_CS1  "MX31865 CS1"
+#define D_SENSOR_MAX31865_CS2  "MX31865 CS2"
+#define D_SENSOR_MAX31865_CS3  "MX31865 CS3"
+#define D_SENSOR_MAX31865_CS4  "MX31865 CS4"
+#define D_SENSOR_MAX31865_CS5  "MX31865 CS5"
```
Comment (arendst) [inline tasmota/language/bg_BG.h:629]: Don't do this. Look at how buttons or relays are implemented. Waste of translation space.

File: tasmota/tasmota_template.h [hunk 2]
```diff
@@ -337,7 +342,8 @@ const char kSensorNames[] PROGMEM =
   D_SENSOR_TELEINFO_RX "|" D_SENSOR_TELEINFO_ENABLE "|"
   D_SENSOR_LMT01_PULSE "|"
   D_SENSOR_IEM3000_TX "|" D_SENSOR_IEM3000_RX "|"
-  D_SENSOR_ZIGBEE_RST
+  D_SENSOR_ZIGBEE_RST "|"
+  D_SENSOR_MAX31865_CS1 "|" D_SENSOR_MAX31865_CS2 "|" D_SENSOR_MAX31865_CS3 "|" D_SENSOR_MAX31865_CS4 "|" D_SENSOR_MAX31865_CS5
```
Comment (arendst) [inline tasmota/tasmota_template.h:346]: See first comment

File: tasmota/tasmota_template_ESP32.h [hunk 3]
```diff
@@ -238,7 +243,8 @@ const char kSensorNames[] PROGMEM =
   D_SENSOR_TELEINFO_RX "|" D_SENSOR_TELEINFO_ENABLE "|"
   D_SENSOR_LMT01_PULSE "|"
   D_SENSOR_IEM3000_TX "|" D_SENSOR_IEM3000_RX "|"
-  D_SENSOR_ZIGBEE_RST
+  D_SENSOR_ZIGBEE_RST "|"
+  D_SENSOR_MAX31865_CS1 "|" D_SENSOR_MAX31865_CS2 "|" D_SENSOR_MAX31865_CS3 "|" D_SENSOR_MAX31865_CS4 "|" D_SENSOR_MAX31865_CS5
```
Comment (arendst) [inline tasmota/tasmota_template_ESP32.h:247]: See previous comment

File: tasmota/tasmota_template_ESP32.h [hunk 4]
```diff
@@ -528,6 +534,13 @@ const uint16_t kGpioNiceList[] PROGMEM = {
   AGPIO(GPIO_MAX31855CLK),    // MAX31855 Serial interface
   AGPIO(GPIO_MAX31855DO),     // MAX31855 Serial interface
 #endif
+#ifdef USE_MAX31865
+  GPIO_SSPI_MAX31865_CS1, //MAX31865 CS selectable pin 1
+  GPIO_SSPI_MAX31865_CS2, //MAX31865 CS selectable pin 2
+  GPIO_SSPI_MAX31865_CS3, //MAX31865 CS selectable pin 3
+  GPIO_SSPI_MAX31865_CS4, //MAX31865 CS selectable pin 4
+  GPIO_SSPI_MAX31865_CS5, //MAX31865 CS selectable pin 5
+#endif
```
Comment (arendst) [inline tasmota/tasmota_template_ESP32.h:543]: Needs AGPIO macro. See jusr before your entries

File: tasmota/xsns_47_max31865.ino [hunk 5]
```diff
@@ -35,71 +35,137 @@
...
+    if (pins_used&(1<<i)) {
+        if (!max31865[i].begin(PTD_WIRES)) {
+        init_status = -1;
+      }
+    }
+  }
 }
 
 /*
 *   MAX31865_GetResult(void)
```
Comment (arendst) [inline tasmota/xsns_47_max31865.ino:122]: too much redundancy.  Good to know: tasmota is all about less code doing more

File: tasmota/xsns_47_max31865.ino [hunk 6]
```diff
@@ -35,71 +35,137 @@
...
-          temperature, resistance, MAX31865_Result.ErrorCode);
+        if(Json){
+          ResponseAppend_P(PSTR(",\"MAX31865(%d)\":{\"" D_JSON_TEMPERATURE "\":%s,\"" D_JSON_RESISTANCE "\":%s,\"" D_JSON_ERROR "\":%d}"), \
+            i, temperature, resistance, MAX31865_Result[i].ErrorCode);
 #ifdef USE_DOMOTICZ
-        if (0 == tele_period) {
-          DomoticzSensor(DZ_TEMP, temperature);
-        }
+          if (0 == tele_period) {
+            DomoticzSensor(DZ_TEMP, temperature);
```
Comment (arendst) [inline tasmota/xsns_47_max31865.ino:152]: Only one temperature can be handled by domoticz and knx. Look at other multisensor implementatioms

File: tasmota/xsns_47_max31865.ino [hunk 7]
```diff
@@ -35,71 +35,137 @@
...
-    dtostrfd(MAX31865_Result.PtdTemp, Settings.flag2.temperature_resolution, temperature);
+        dtostrfd(MAX31865_Result[i].PtdResistance, Settings.flag2.temperature_resolution, resistance);
+        dtostrfd(MAX31865_Result[i].PtdTemp, Settings.flag2.temperature_resolution, temperature);
 
-    if(Json){
-        ResponseAppend_P(PSTR(",\"MAX31865\":{\"" D_JSON_TEMPERATURE "\":%s,\"" D_JSON_RESISTANCE "\":%s,\"" D_JSON_ERROR "\":%d}"), \
-          temperature, resistance, MAX31865_Result.ErrorCode);
+        if(Json){
+          ResponseAppend_P(PSTR(",\"MAX31865(%d)\":{\"" D_JSON_TEMPERATURE "\":%s,\"" D_JSON_RESISTANCE "\":%s,\"" D_JSON_ERROR "\":%d}"), \
+            i, temperature, resistance, MAX31865_Result[i].ErrorCode);
```
Comment (arendst) [inline tasmota/xsns_47_max31865.ino:149]: Do not use brackets a sindex. Lok how other multisensor drivers are doing it like bh1750 allowing either a dash or underline as index.

**Comment by arendst** [comment]
As the total amount of gpios is nearing 250.as being the max on legacy esp8266 I'm reluctant to add 5 new ones while the usecase is specific you. I'm more thinking of reusing gpios defined as switch. I'll need to dive in if this is even possible.

**Comment by brvdboss** [comment]
Thanks for all the feedback!  I'll improve the pull request based on your feedback and by looking at the other drivers you referenced. I will likely only be able to do this next weekend

**Comment by arendst** [comment]
Stay open.

**Comment by arendst** [comment]
Thx. Merged with changes.

**Comment by kobusdbr** [comment]
Hello all. I am working with a brewery here in south Africa. They use PT100 temperature sensors in their fermenters. I would like to add a D1 Mini to each fermenter so that they can read and display temperature data. Want to use HA.  We will then use those sensors to switch relays to switch on solenoids to cool the brew. The Home assistant thermostat will work for this. I have found some info on using PT100 with arduino but nothing for tasmota as yet. Has anyone got some experience with this that are able to point me in the right direction?

**Comment by lalo-uy** [comment]
It is suported, but not included in default bin.
You will need to compile yourself with

define USE_MAX31865

**Comment by bryanb334** [comment]
I see the max31855 listed in the sensor bin.  If I want to use muntiple max31855/thermocouples so I still need to complile my own binary?  Thanks

**Comment by barbudor** [comment]
You are confusing MAX31855 and MAX31865
The first one is indeed included in the sensor bin but support only 1 device
The 2nd one (the one this PR is about) is not included by default and support up to 6 devices.

**Comment by bryanb334** [comment]
Thanks for your reply.

**Pull request closed without merging**